### ISSUE TITLE
Chapter4 - Finetuning and Caltech256

### ISSUE THREAD

**Issue by jhagege** [PracticalDL/Practical-Deep-Learning-Book#44]
Hi, do you happen to have the code for finetuning Resnet on Caltech 101, and the code for extracting on calnet256 ? 
Thanks !

**Comment by jhagege** [comment]
By the way, do you recommend freezing the conv layers when doing fine-tuning in this case ? 
Or train on all the layers (conv + FC) ? 
Thanks.

**Comment by jhagege** [comment]
Well, when removing the following lines:
```
#     for layer in base_model.layers[:]:
#         layer.trainable = False
```
I get a much lower validation accuracy (0.7363 compared to 0.85) when finetuning Caltech101, so I would suggest putting back those lines in the code sample in chapter 4 after:
> To compare the process of fine tuning a model optimized for classification tasks as opposed to similarity search, let’s recall how we fine tuned our model in Chapter 3 for classification:
My 2 cents :)

**Comment by jhagege** [comment]
I tried running calculate_accuracy(feature_list_compressed[:]) on caltech256 with finetuned Resnet, I didn't get better accuracy than 69.69 using PCA. 
On your table with PCA + brute force you report 95.34.
Can you provide hints / code regarding how you reached that good a score ? 
Thanks much !

**Comment by sidgan** [comment]
> Hi, do you happen to have the code for finetuning Resnet on Caltech 101, and the code for extracting on calnet256 ?
> Thanks !

Refer to the finetuning pipeline in [`chapter-3/1-keras-custom-classifier-with-transfer-learning.ipynb`](https://github.com/PracticalDL/Practical-Deep-Learning-Book/blob/master/code/chapter-3/1-keras-custom-classifier-with-transfer-learning.ipynb). You can read the Caltech101 dataset according to the code provided, just remember to update the Cats and Dogs dataset with Caltech101. 
Once you have the trained model, use the code in [`chapter-4/1-feature-extraction.ipynb`](https://github.com/PracticalDL/Practical-Deep-Learning-Book/blob/master/code/chapter-4/1-feature-extraction.ipynb) to extract the features.

**Comment by sidgan** [comment]
> By the way, do you recommend freezing the conv layers when doing fine-tuning in this case ?
> Or train on all the layers (conv + FC) ?
> Thanks.

Fine-tuning a few layers or the entire model depends on how large the dataset at hand is, and how similar it is to the original dataset the model was trained on. The more data you have, the more layers you are able to unfreeze. If you are training the ResNet50 model on the Caltech101 dataset you should be able to get a high enough accuracy with finetuning just the final few layers or training on the embeddings as demonstrated in [Chapter-4](https://github.com/PracticalDL/Practical-Deep-Learning-Book/tree/master/code/chapter-4)

**Comment by sidgan** [comment]
> Well, when removing the following lines:
> 
> ```
> #     for layer in base_model.layers[:]:
> #         layer.trainable = False
> ```
> 
> I get a much lower validation accuracy (0.7363 compared to 0.89) when finetuning Caltech101, so I would suggest putting back those lines in the code sample in chapter 4 after:
> 
> > To compare the process of fine tuning a model optimized for classification tasks as opposed to similarity search, let’s recall how we fine tuned our model in Chapter 3 for classification:
> > My 2 cents :)

That's a good point. We will have to update the errata.

**Comment by sidgan** [comment]
> I tried running calculate_accuracy(feature_list_compressed[:]) on caltech256 with finetuned Resnet, I didn't get better accuracy than 69.69 using PCA.
> On your table with PCA + brute force you report 95.34.
> Can you provide hints / code regarding how you reached that good a score ?
> Thanks much !

When we were running the code, we looked at the parameters such as `num_feature_dimensions` and played around to reflect the best accuracy possible. The value of `num_feature_dimensions = 100` gave us the best accuracy and as such, we kept it in the code. 

Just to verify - did you happen to edit the feature length, the number of images in Caltech256, model, or any other variables in [`chapter-4/1-feature-extraction.ipynb`](https://github.com/PracticalDL/Practical-Deep-Learning-Book/blob/master/code/chapter-4/1-feature-extraction.ipynb)?

**Comment by sidgan** [comment]
> > Hi, do you happen to have the code for finetuning Resnet on Caltech 101, and the code for extracting on calnet256 ?
> > Thanks !
> 
> Refer to the finetuning pipeline in [`chapter-3/1-keras-custom-classifier-with-transfer-learning.ipynb`](https://github.com/PracticalDL/Practical-Deep-Learning-Book/blob/master/code/chapter-3/1-keras-custom-classifier-with-transfer-learning.ipynb). You can read the Caltech101 dataset according to the code provided, just remember to update the Cats and Dogs dataset with Caltech101.
> Once you have the trained model, use the code in [`chapter-4/1-feature-extraction.ipynb`](https://github.com/PracticalDL/Practical-Deep-Learning-Book/blob/master/code/chapter-4/1-feature-extraction.ipynb) to extract the features.

Hi @jhagege we have updated the [chapter-4/1-feature-extraction.ipynb](https://github.com/PracticalDL/Practical-Deep-Learning-Book/blob/master/code/chapter-4/1-feature-extraction.ipynb) notebook to reflect the finetuning code.

**Comment by jhagege** [comment]
Great, thanks much !
